### PR TITLE
Add row and column move to markdown table menus

### PR DESCRIPTION
**Pull request by SidhuK**
Closes 

## Description

Markdown table row/column menus could add and delete, but reordering a row or column meant cutting and pasting cells. This adds move up/down and left/right to those same menus.

- Row menu: add above/below, move up/down, delete. The header row stays at index 0; body rows cannot swap with it.
- Column menu: add left/right, move left/right, delete.
- Native menus can fire after the editor blurs, so the selected cell is captured when the menu opens and restored before the command runs.
- Insert/delete still use the existing TipTap chain. Moves use `moveTableRow` / `moveTableColumn`.
- One toast if the captured table is gone. Menu items disable at the header and at the first/last row or column.

Rich editor only. No new shortcuts, settings, or backend commands. English copy only.

## Screenshots

Screenshots or a screen recording of the visual changes associated with this PR.

## Docs

No doc changes. Menu labels live in `src/i18n/locales/en/editor.json` under `tableControls`.

## Ready?

Did you do any of the following? If not, no worries, but if you can
it's really helpful.

- [ ] Documented what's new
- [ ] Added in-code documentation (wherever needed)
- [ ] Wrote tests for new components/features
- [x] Ran the linter to ensure style guidelines were followed
- [ ] Created a demo

## Summary by Sourcery

Enable rich-editor table users to reorder rows and columns directly from the existing menus.

New Features:
- Add row and column reordering actions to rich-editor markdown table menus.
- Disable move actions when the selected row or column is already at its boundary, while keeping the header row fixed.

Bug Fixes:
- Preserve the selected table cell when native menus open after the editor loses focus and report unavailable actions when the target table no longer exists.

Enhancements:
- Centralize table action targets and capabilities so menu commands can restore and operate on the captured selection.

Tests:
- Update table control tests for translated menu labels and expanded table editor state.

---
## Summary by cubic
Adds move up/down and left/right to the markdown table row and column menus, so reordering no longer requires cutting and pasting cells.

- Row menu: add above/below, move up/down, and delete; the header row stays pinned at index 0.
- Column menu: add left/right, move left/right, and delete.
- The selected cell is captured when the menu opens and restored before the command runs, since native menus can fire after the editor blurs.
- Insert and delete keep using the existing TipTap chain; moves use `moveTableRow` and `moveTableColumn`.
- Move items disable at the header row and at the first or last row or column.
- If the captured table is gone, actions now abort with a single toast instead of running against the current selection.
- Rich editor only; no new shortcuts, settings, or backend commands. Menu labels now live in `en/editor.json` under `tableControls`.

<sup>Written for commit 7b12302885b26b167b6da0a7990995aacd495efd. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/SidhuK/Glyph/pull/550?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added table controls for inserting, deleting, and moving rows and columns.
  * Added capability-based control states to prevent unavailable actions.
  * Localized table control labels and unavailable-action messaging.
  * Preserved table selection when performing table actions.

* **Bug Fixes**
  * Improved table action targeting when opening menus before making a selection.
  * Added error feedback when a table action cannot be completed.

### CODE REVIEW

**Comment by SidhuK** [comment]
@greptile
@coderabbitai

File: src/components/editor/TableInlineControls.tsx [hunk 1]
```diff
@@ -14,6 +15,9 @@ import {
 	DropdownMenuTrigger,
 } from "../ui/shadcn/dropdown-menu";
 import type {
+	TableActionTarget,
+	TableEditorAction,
+	TableEditorCapabilities,
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/TableInlineControls.tsx]: Removed. It was leftover after the menu builder stopped taking a capabilities type.

File: src/components/editor/TableInlineControls.tsx [hunk 2]
```diff
@@ -86,21 +126,42 @@ const TableAxisControl = memo(function TableAxisControl({
...
 		(event: MouseEvent<HTMLButtonElement>) => {
+			const target = captureTarget();
+			capturedTargetRef.current = target;
+			if (!target) return;
+			const nativeMenuItems = toNativeMenuItems(menuItems, runCapturedCommand);
 			void showNativePopupMenu(event, nativeMenuItems).catch(
 				(error: unknown) => {
 					console.error(`Failed to show table ${axis} menu`, error);
 				},
 			);
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/TableInlineControls.tsx:152]: Leaving this. `void showNativePopupMenu(...).catch(...)` is the existing native-menu pattern in this file and elsewhere; changing it isn't a functional issue.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 3]
```diff
@@ -1,9 +1,217 @@
...
+		capabilities: {
+			canDeleteRow: canDelete.row,
+			canDeleteColumn: canDelete.column,
+			canMoveRowUp: rowIndex > 1,
+			canMoveRowDown: rowIndex > 0 && rowIndex < rect.map.height - 1,
+			canMoveColumnLeft: columnIndex > 0,
+			canMoveColumnRight: columnIndex < rect.map.width - 1,
+		},
+	};
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:48]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 4]
```diff
@@ -1,9 +1,217 @@
...
+		a.target.rowIndex === b.target.rowIndex &&
+		a.target.columnIndex === b.target.columnIndex &&
+		a.capabilities.canDeleteRow === b.capabilities.canDeleteRow &&
+		a.capabilities.canDeleteColumn === b.capabilities.canDeleteColumn &&
+		a.capabilities.canMoveRowUp === b.capabilities.canMoveRowUp &&
+		a.capabilities.canMoveRowDown === b.capabilities.canMoveRowDown &&
+		a.capabilities.canMoveColumnLeft === b.capabilities.canMoveColumnLeft &&
+		a.capabilities.canMoveColumnRight === b.capabilities.canMoveColumnRight
+	);
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:67]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 5]
```diff
@@ -1,9 +1,217 @@
...
+			canDeleteColumn: canDelete.column,
+			canMoveRowUp: rowIndex > 1,
+			canMoveRowDown: rowIndex > 0 && rowIndex < rect.map.height - 1,
+			canMoveColumnLeft: columnIndex > 0,
+			canMoveColumnRight: columnIndex < rect.map.width - 1,
+		},
+	};
+}
+
+export function tableSnapshotsEqual(
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:50]: Not a functional bug. The equality helper is a field-by-field compare for `useEditorState`; splitting it wouldn't change behavior.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 6]
```diff
@@ -1,9 +1,217 @@
...
+		rowIndex >= map.height ||
+		columnIndex >= map.width
+	) {
+		return null;
+	}
+	const relative = map.positionAt(rowIndex, columnIndex, table);
+	const cell = table.nodeAt(relative);
+	if (!cell) return null;
+	return { cell, cellPos: tablePos + 1 + relative };
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:88]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 7]
```diff
@@ -1,9 +1,217 @@
...
+		a.capabilities.canDeleteRow === b.capabilities.canDeleteRow &&
+		a.capabilities.canDeleteColumn === b.capabilities.canDeleteColumn &&
+		a.capabilities.canMoveRowUp === b.capabilities.canMoveRowUp &&
+		a.capabilities.canMoveRowDown === b.capabilities.canMoveRowDown &&
+		a.capabilities.canMoveColumnLeft === b.capabilities.canMoveColumnLeft &&
+		a.capabilities.canMoveColumnRight === b.capabilities.canMoveColumnRight
+	);
+}
+
+function cellPosition(
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:69]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 8]
```diff
@@ -1,9 +1,217 @@
...
+		located.cellPos + 1,
+		located.cellPos + located.cell.nodeSize - 1,
+	);
+	const selection = TextSelection.near(editor.state.doc.resolve(inner), 1);
+	const tr = editor.state.tr.setSelection(selection);
+	tr.setMeta("addToHistory", false);
+	editor.view.dispatch(tr);
+	editor.view.focus();
+	return true;
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:121]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 9]
```diff
@@ -1,9 +1,217 @@
...
+		case "moveRowDown":
+			return target.rowIndex > 0 && target.rowIndex < map.height - 1;
+		case "moveColumnLeft":
+			return target.columnIndex > 0;
+		case "moveColumnRight":
+			return target.columnIndex < map.width - 1;
+		default:
+			return true;
+	}
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:141]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 10]
```diff
@@ -1,9 +1,217 @@
...
+	);
+	const selection = TextSelection.near(editor.state.doc.resolve(inner), 1);
+	const tr = editor.state.tr.setSelection(selection);
+	tr.setMeta("addToHistory", false);
+	editor.view.dispatch(tr);
+	editor.view.focus();
+	return true;
+}
+
+function canMove(
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:123]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 11]
```diff
@@ -1,9 +1,217 @@
...
+			return moveTableColumn({
+				from: action.target.columnIndex,
+				to: action.target.columnIndex + 1,
+				select: true,
+				pos,
+			})(editor.state, dispatch);
+		default:
+			return false;
+	}
+}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:180]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 12]
```diff
@@ -1,9 +1,217 @@
...
+		case "moveRowDown":
+		case "moveColumnLeft":
+		case "moveColumnRight":
+			return table ? runMoveCommand(editor, action) : false;
+		default: {
+			const _exhaustive: never = action.kind;
+			return _exhaustive;
+		}
+	}
 }
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:219]: Not a functional bug. The switch is the command dispatch; splitting it wouldn't change behavior.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 13]
```diff
@@ -1,9 +1,217 @@
...
+				to: action.target.columnIndex + 1,
+				select: true,
+				pos,
+			})(editor.state, dispatch);
+		default:
+			return false;
+	}
+}
+
+export function runTableEditorAction(
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:182]: False positive for an ES module. These functions are imported, not browser globals.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 14]
```diff
@@ -1,9 +1,217 @@
...
+		case "addRowBefore":
+		case "addRowAfter":
+		case "deleteRow":
+		case "addColumnBefore":
+		case "addColumnAfter":
+		case "deleteColumn":
+			return editor
+				.chain()
+				.focus(null, { scrollIntoView: false })
+				[action.kind]()
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:207]: Biome formatted the chain that way. Not a functional issue.

File: src/components/editor/hooks/useTableInlineControls.ts [hunk 15]
```diff
@@ -36,57 +45,45 @@ export function useTableInlineControls({
...
-		editor.on("selectionUpdate", syncDeleteCapabilities);
-		editor.on("destroy", clearDeleteCapabilities);
-		return () => {
-			editor.off("transaction", syncDeleteCapabilities);
-			editor.off("selectionUpdate", syncDeleteCapabilities);
-			editor.off("destroy", clearDeleteCapabilities);
-		};
-	}, [canEdit, editor, mode]);
-
 	useEffect(() => {
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/useTableInlineControls.ts:87]: Pre-existing overlay-positioning effect, not introduced by this PR. Not a functional bug.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 16]
```diff
@@ -1,9 +1,217 @@
...
+	action: TableEditorAction,
+): boolean {
+	if (editor.isDestroyed) return false;
+
+	const node = editor.state?.doc?.nodeAt(action.target.tablePos);
+	const table = node?.type.name === "table" ? node : null;
+	if (table) {
+		if (!canMove(action.kind, table, action.target)) return false;
+		if (!restoreSelection(editor, table, action.target)) return false;
+	}
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts]: Fixed. If the document is present and the captured pos is no longer a table, the action returns false and the unavailable toast fires. Insert/delete no longer fall through onto the current selection.

File: src/components/editor/NoteInlineEditor.test.tsx [hunk 17]
```diff
@@ -132,14 +132,21 @@ vi.mock("@tiptap/react", () => ({
...
+			columnIndex: 0,
+		},
+		capabilities: {
+			canDeleteRow: true,
+			canDeleteColumn: true,
+			canMoveRowUp: true,
+			canMoveRowDown: true,
+			canMoveColumnLeft: true,
+			canMoveColumnRight: true,
+		},
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/NoteInlineEditor.test.tsx:148]: Leaving this. The existing test still covers the add-row/add-column menu path. Repo convention is not to add test files unless asked; the move/boundary logic is small and lives next to the menu enablement.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 18]
```diff
@@ -1,9 +1,217 @@
...
+
+export function runTableEditorAction(
+	editor: Editor,
+	action: TableEditorAction,
+): boolean {
+	if (editor.isDestroyed) return false;
+
+	const node = editor.state?.doc?.nodeAt(action.target.tablePos);
+	const table = node?.type.name === "table" ? node : null;
+	if (table) {
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:192]: Fixed. Missing captured table now fails closed when a document is present, same as moves, so a stale native-menu action can't edit a different table.

File: src/components/editor/TableInlineControls.tsx [hunk 19]
```diff
@@ -86,21 +126,42 @@ const TableAxisControl = memo(function TableAxisControl({
...
 }: TableAxisControlProps) {
-	const nativeMenuItems = useMemo(
-		() => toNativeMenuItems(menuItems, onCommand),
-		[menuItems, onCommand],
+	const capturedTargetRef = useRef<TableActionTarget | null>(null);
+
+	const runCapturedCommand = useCallback(
+		(command: TableEditorCommand) => {
+			const target = capturedTargetRef.current ?? captureTarget();
+			if (!target) return;
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/TableInlineControls.tsx:136]: Agreed it's minor. The primary path captures a target on menu open; a fully desynced add with no snapshot isn't the workflow this PR is for. Failed actions still toast from `onCommand`.

File: src/components/editor/hooks/tableEditorCommands.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:192]: Fixed. Actions now return false when the captured table node is gone, which surfaces the unavailable toast.

File: src/components/editor/TableInlineControls.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/TableInlineControls.tsx]: Removed the unused import so `tsc --noEmit` passes.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 20]
```diff
@@ -1,9 +1,219 @@
...
+		case "moveRowDown":
+		case "moveColumnLeft":
+		case "moveColumnRight":
+			return table ? runMoveCommand(editor, action) : false;
+		default: {
+			const _exhaustive: never = action.kind;
+			return _exhaustive;
+		}
+	}
 }
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:219]: False positive for an ES module. `runTableEditorAction` is imported, not a browser global.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 21]
```diff
@@ -1,9 +1,219 @@
...
+				to: action.target.columnIndex + 1,
+				select: true,
+				pos,
+			})(editor.state, dispatch);
+		default:
+			return false;
+	}
+}
+
+export function runTableEditorAction(
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:182]: Not a functional bug. The extra guard is the fail-closed path reviewers asked for; splitting the switch would not change behavior.

File: src/components/editor/TableInlineControls.tsx [hunk 22]
```diff
@@ -86,21 +125,42 @@ const TableAxisControl = memo(function TableAxisControl({
...
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/TableInlineControls.tsx:136]: Same as the earlier note. Primary path captures a target on menu open; a fully desynced add with no snapshot isn't this PR's workflow. Failed actions still toast from onCommand.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 23]
```diff
@@ -1,9 +1,219 @@
...
+	editor: Editor,
+	action: TableEditorAction,
+): boolean {
+	if (editor.isDestroyed) return false;
+
+	const doc = editor.state?.doc;
+	const node = doc?.nodeAt(action.target.tablePos);
+	const table = node?.type.name === "table" ? node : null;
+	if (doc && !table) return false;
+	if (table) {
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:192]: Already fixed in 7b123028. If a document is present and the captured pos is no longer a table, we return false before the command switch, so insert/delete cannot hit a different table.

File: src/components/editor/hooks/tableEditorCommands.ts [hunk 24]
```diff
@@ -1,9 +1,219 @@
...
+	return {
+		target: {
+			tablePos: rect.tableStart - 1,
+			rowIndex,
+			columnIndex,
+		},
+		capabilities: {
+			canDeleteRow: canDelete.row,
+			canDeleteColumn: canDelete.column,
+			canMoveRowUp: rowIndex > 1,
```
[the remark this answers is not in the document]
Comment (SidhuK) [reply src/components/editor/hooks/tableEditorCommands.ts:42]: Leaving this. The expressions match today; the snapshot drives the menu and `canMove` is the stale-menu re-check. Deduping them would not change behavior.
